External/FEXCore/Source/Interface/Context/Context.h

```diff
@@ -173,6 +173,11 @@ namespace FEXCore::Context {
       RemoveThreadCodeEntry(Frame->Thread, GuestRIP);
     }
 
+    // returns false if a handler was already registered
```

Comment:
> Are we actually handling that scenario properly currently?

It just returns false (and does nothing). It's up to the caller to recover, or assert.

> For guest-callable host function pointers in libGL/libvulkan, the only thing we need is idempotence: Calling AddCustomIREntrypoint twice with the same Entrypoint/Handler values would be fine, but linking the same Entrypoint to two different Handlers would be a fatal error.

That was my thinking, that's why I added the return value instead of assert.

(Thinking on it, as we don't keep track of the actual set values (captured in the lambda) it'll need to keep a parallel map there or... something?)

Reply:
> It just returns false (and does nothing). It's up to the caller to recover, or assert.
> That was my thinking, that's why I added the return value instead of assert.

This shifts the need for reliable error handling from the implementation to the caller but doesn't obviate the need.

For libGL specifically we can't do reliable error handling, since in glXGetProcAddress we can't check if the function name has already been queried before (=> safe to ignore error) or something else unexpectedly linked this function pointer to custom IR already (=> fatal error). The alternative is to ignore errors, risking to miss pointer confusion bugs (and begging the question of what use the returned error information was in the first place).

> (Thinking on it, as we don't keep track of the actual set values (captured in the lambda) it'll need to keep a parallel map there or... something?)

Extending the interface by an actual type (say `IRGenerator`) in place of the `std::function` makes the code more expressive while also allowing to address this issue without having to track external state. Some trickery needs to be added to compare across different subclasses of IRGenerator without RTTI, though. Here's one possible approach:
```cpp
struct IRGenerator {
    uint32_t typeid;

    virtual void(uintptr_t Entrypoint, FEXCore::IR::IREmitter *) = 0;

    bool operator==(const IRGenerator& oth) const {
        return typeid == oth.typeid && Compare(oth);
    }

private:
    // Type-specific comparison
    virtual bool Compare(const IRGenerator&) = 0;
};

struct LibGLIRGenerator : IRGenerator {
    // typeid = 0x12341234 in constructor

    uint32_t GuestThunkEntrypoint;

    void(uintptr_t Entrypoint, FEXCore::IR::IREmitter *) override;

    bool Compare(const LibGLIRGenerator&) override {
        return GuestThunkEntrypoint == oth.GuestThunkEntrypoint;
    }
};

std::unordered_map<uint64_t, std::unique_ptr<IRGenerator>> CustomIRHandlers;
```

Reply:
Ah, no, it is passed by reference. I'd just pass another tag or something.

Reply:
> You can just pass a function subclass that overloads the operator() there, can't you?

Are you suggesting to inherit from `std::function`? I'm not sure that's supported, and it doesn't buy you anything over using `unique_ptr` and virtual functions.

> Though, that'd still need a getter. Is typeid available without RTTI?

Nope, hence the `typeid` dance above. Then again we appear not to actually disable RTTI in FEX, so we might as well just use it here?

Reply:
> doesn't buy you anything over using unique_ptr and virtual functions.

Having to create classes manually for simple things is pretty ugly

>  so we might as well just use it here?

I don't think that is wise on our side. We are already too coupled to the CRT/STL, and there are a few known reasons that we'll have to decouple in the future.

Honestly, I'd just keep a second map, or add a second validation function / taging void * / whatever.

Reply:
> I don't think that is wise on our side. We are already too coupled to the CRT/STL, and there are a few known reasons that we'll have to decouple in the future.

Isn't RTTI just a bunch of extra information added to the vtable? I'm not an expert on the challenges which might be involved here, but sounds like if RTTI were to become a problem, we'd also have to avoid using virtual functions in the first place (which also means avoiding things like `std::function`).

> Honestly, I'd just add a second map, or add a second validation function / taging void * / whatever. There's no reason to over engineer this.

That is not a feasible road forward for thunked vtables. We can't add cumbersome error checking involving external state tracking in every single function that returns an object with function pointers.

An alternative is to use a single `class IRGenerator` for everything (without subclasses) and then just unconditionally have the GuestThunkEntrypoint as an optionally-used member. That design requires more changes if new requirements come up, but should suffice for the time being.

Reply:
> We can just add a void *Setter, void *Data to the adder, and a getter. Then we can do
> 
> if (Setter == this) { // treat Data accordingly, eg as a uintptr_t for the common case, or a pointer to a more complex structure
> }
> 
> Then we can just unique_ptr lambda capture, and have the data clean itself when the object gets destroyed.
> 
> Also, this entire operation needs to be interlocked, so a lock needs to be returned as well.

I'm not following what exactly you're proposing. Do you want to elaborate or should I wait for an implementation of this proposal?

FWIW `std::any` might come in handy for what you're proposing, though at the same time this all sounds like reinventing an ad-hoc version of inheritance.

Reply:
Is that ok?

Reply:
(rebased `feat/gch-trampolines`)

Reply:
I don't think the actual requirements we have justify this overly convoluted design.

Why does `Data` need to be an arbitrary `void*` pointer? We only need it in one specific context currently, so we might as well bake that into the design for now. Changing `Data` to be the `uintptr_t` used for GCH allows to move the error handling into AddCustomIREntrypoint and to drop the lock handling.

Reply:
This is also used for the IR tests, and might have other uses in the future - I don't want to bake in the GCH logic in the core.